ext/objspace/objspace_dump.c

```diff
@@ -37,22 +40,104 @@ struct dump_config {
     unsigned int full_heap: 1;
 };
 
-PRINTF_ARGS(static void dump_append(struct dump_config *, const char *, ...), 2, 3);
 static void
-dump_append(struct dump_config *dc, const char *format, ...)
+dump_flush(struct dump_config *dc)
 {
-    va_list vl;
+    if (dc->buffer_len) {
+        if (dc->stream) {
+            fwrite(dc->buffer, 1, dc->buffer_len, dc->stream);
+        }
+        else if (dc->string) {
+            rb_str_cat(dc->string, dc->buffer, dc->buffer_len);
+        }
+        dc->buffer_len = 0;
+    }
+}
+
+static inline void
+buffer_inc_capa(struct dump_config *dc, unsigned long requested)
+{
+    unsigned long required;
+
+    if (requested + dc->buffer_len >= dc->buffer_capa) {
+        dump_flush(dc);
+    }
+
+    for (required = dc->buffer_capa; requested > required - dc->buffer_len; required <<= 1);
+
+    if (required > dc->buffer_capa) {
+        REALLOC_N(dc->buffer, char, required);
+        dc->buffer_capa = required;
+    }
+}
+
+static void buffer_append(struct dump_config *dc, const char *cstr, unsigned long len)
+{
+    if (len > 0) {
+        buffer_inc_capa(dc, len);
+        MEMCPY(dc->buffer + dc->buffer_len, cstr, char, len);
+        dc->buffer_len += len;
+    }
+}
+
+# define dump_append(dc, str) buffer_append(dc, (str), (long)strlen(str))
+
+PRINTF_ARGS(static void dump_appendf(struct dump_config *, const char *, ...), 2, 3);
+static void
+dump_appendf(struct dump_config *dc, const char *format, ...)
+{
+    va_list vl, vl2;
     va_start(vl, format);
+    va_copy(vl2, vl);
+
+    unsigned long capa_left = dc->buffer_capa - dc->buffer_len;
+    unsigned long required = vsnprintf(dc->buffer + dc->buffer_len, capa_left, format, vl);
 
-    if (dc->stream) {
-	vfprintf(dc->stream, format, vl);
+    if (required >= capa_left) {
+        buffer_inc_capa(dc, required + 1);
+        capa_left = dc->buffer_capa - dc->buffer_len;
+        vsnprintf(dc->buffer + dc->buffer_len, capa_left, format, vl2);
     }
-    else if (dc->string)
-	rb_str_vcatf(dc->string, format, vl);
+    dc->buffer_len += required;
 
     va_end(vl);
 }
 
+static void
+dump_appendc(struct dump_config *dc, char c)
+{
+    buffer_inc_capa(dc, 1);
+    dc->buffer[dc->buffer_len] = c;
+    dc->buffer_len++;
+}
+
+static const char hex_table[16] = {
+  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
+};
```

Comment:
```suggestion
static const char *const hex_table = "0123456789abcdef";
```

It seems this style is used throughout the project (try `git grep "0123456789abcdef"`).

Reply:
See `ruby_hexdigits` and `ruby_digitmap`.
They are shared and used more.